src/bbs2gh/Services/BbsSmbArchiveDownloader.cs

```diff
@@ -0,0 +1,227 @@
+
+    public async Task<string> Download(long exportJobId, string targetDirectory = IBbsArchiveDownloader.DEFAULT_TARGET_DIRECTORY)
+    {
+        _nextProgressReport = DateTime.Now;
+
+        ISMBFileStore fileStore = null;
+        object sourceExportArchiveHandle = null;
+
+        var sourceExportArchiveFullPath = GetSourceExportArchiveAbsolutePath(exportJobId);
+        var share = sourceExportArchiveFullPath[..sourceExportArchiveFullPath.IndexOf("\\", StringComparison.Ordinal)];
+        var sourceExportArchivePathAfterShare = sourceExportArchiveFullPath[(sourceExportArchiveFullPath.IndexOf("\\", StringComparison.Ordinal) + 1)..];
+
+        var targetExportArchiveFullPath =
+            Path.Join(targetDirectory ?? IBbsArchiveDownloader.DEFAULT_TARGET_DIRECTORY, IBbsArchiveDownloader.GetExportArchiveFileName(exportJobId)).ToUnixPath();
+
+        await using var targetExportArchive = OpenWriteTargetExportArchive(targetExportArchiveFullPath);
+
+        try
+        {
+            ConnectToHost();
+            Login();
+
+            fileStore = CreateSmbFileStore(share);
+            sourceExportArchiveHandle = CreateFileHandle(fileStore, sourceExportArchivePathAfterShare);
+            var sourceExportArchiveSize = GetFileSize(fileStore, sourceExportArchiveHandle);
+
+            long bytesRead = 0;
+            while (true)
+            {
+                var status = fileStore.ReadFile(out var data, sourceExportArchiveHandle, bytesRead, (int)_smbClient.MaxReadSize);
+
+                if (IsEndOfFileStatus(status) || data.Length == 0)
+                {
+                    break;
+                }
+
+                if (!IsSuccessStatus(status))
+                {
+                    throw new OctoshiftCliException($"Failed to read from source export archive \"{sourceExportArchiveFullPath}\" (Status Code: {status}).");
+                }
+
+                bytesRead += data.Length;
+                await _fileSystemProvider.WriteAsync(targetExportArchive, data);
+
+                LogProgress(bytesRead, sourceExportArchiveSize);
+            }
+
+            return targetExportArchiveFullPath;
+        }
+        finally
+        {
+            if (sourceExportArchiveHandle != null)
+            {
+                fileStore?.CloseFile(sourceExportArchiveHandle);
+            }
+
+            fileStore?.Disconnect();
+            _smbClient.Logoff();
+            _smbClient.Disconnect();
+        }
+    }
+
+    private void LogProgress(long downloadedBytes, long? totalBytes)
+    {
+        if (DateTime.Now < _nextProgressReport)
+        {
+            return;
+        }
+
+        var totalProgressMessage = totalBytes.HasValue
+            ? $" out of {GetLogFriendlySize(totalBytes.Value)} ({GetPercentage(downloadedBytes, totalBytes.Value)})"
+            : "";
+        _log.LogInformation($"Download archive in progress, {GetLogFriendlySize(downloadedBytes)}{totalProgressMessage} completed...");
```

Comment:
```suggestion
        _log.LogInformation($"Archive download in progress, {GetLogFriendlySize(downloadedBytes)}{totalProgressMessage} completed...");
```

Reply:
this also needs to be applied to SSH downloader too.

Reply:
addressed in 28ca1022e7e2599f3e374aa83643ce8dd535f1e2